File: sesion2-lab/aspnet-identity/identity-demo/Controllers/ApiController.cs
using System.Security.Claims;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace IdentityDemo.Controllers;

[ApiController]
[Route("api")]
public class ApiController : ControllerBase
{
    [HttpGet("public/hello")]
    [AllowAnonymous]
    public IActionResult PublicHello()
    {
        return Ok(new
        {
            message = "Hola! Este endpoint es publico, no requiere token.",
            scope = "public"
        });
    }

    [HttpGet("me")]
    [Authorize]
    public IActionResult Me()
    {
        var username = User.FindFirstValue("preferred_username")
            ?? User.FindFirstValue(ClaimTypes.Name)
            ?? User.Identity?.Name;
        var email = User.FindFirstValue(ClaimTypes.Email);
        var authorities = User.FindAll(ClaimTypes.Role).Select(c => "ROLE_" + c.Value).ToList();

        return Ok(new
        {
            message = "Estas autenticado.",
            user = username,
            email,
            authorities
        });
    }

    [HttpGet("admin/hello")]
    [Authorize(Roles = "Admin")]
    public IActionResult AdminHello()
    {
        var username = User.FindFirstValue("preferred_username")
            ?? User.FindFirstValue(ClaimTypes.Name)
            ?? User.Identity?.Name;

        return Ok(new
        {
            message = "Bienvenido al area de administracion.",
            user = username,
            scope = "admin"
        });
    }
}

File: sesion2-lab/entra-spring-security/resource-server/src/main/java/com/example/demo/web/ApiController.java
package com.example.demo.web;

import java.util.List;
import java.util.Map;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api")
public class ApiController {

    @GetMapping("/public/hello")
    public Map<String, Object> publicHello() {
        return Map.of(
            "message", "Hola! Este endpoint es publico, no requiere token.",
            "scope", "public",
            "idp", "Microsoft Entra ID");
    }

    @GetMapping("/me")
    public Map<String, Object> me(@AuthenticationPrincipal Jwt jwt, Authentication authentication) {
        List<String> authorities = authentication.getAuthorities().stream()
            .map(GrantedAuthority::getAuthority)
            .toList();
        String user = firstNonBlank(
            jwt.getClaimAsString("preferred_username"),
            jwt.getClaimAsString("upn"),
            jwt.getClaimAsString("name"),
            jwt.getSubject());
        return Map.of(
            "message", "Estas autenticado.",
            "user", user,
            "email", String.valueOf(jwt.getClaimAsString("email")),
            "authorities", authorities,
            "roles", jwt.getClaimAsStringList("roles") != null
                ? jwt.getClaimAsStringList("roles")
                : List.of());
    }

    @GetMapping("/admin/hello")
    public Map<String, Object> adminHello(@AuthenticationPrincipal Jwt jwt) {
        String user = firstNonBlank(
            jwt.getClaimAsString("preferred_username"),
            jwt.getClaimAsString("upn"),
            jwt.getClaimAsString("name"),
            jwt.getSubject());
        return Map.of(
            "message", "Bienvenido al area de administracion.",
            "user", user,
            "scope", "admin");
    }

    private static String firstNonBlank(String... values) {
        for (String value : values) {
            if (value != null && !value.isBlank()) {
                return value;
            }
        }
        return "";
    }
}
